### ISSUE TITLE
Session: paper overview

### ISSUE THREAD

**Issue by gar1t** [gar1t/tensorflow-workshop#7]
We should take some time, probably the first session, to present the original paper:

https://arxiv.org/abs/1703.06870

Ideally we'd have someone with background in the data science/mathematics side who can answer some of the harder questions that might come up.

**Comment by gar1t** [comment]
Presentation on selective search:

http://www.cs.cornell.edu/courses/cs7670/2014sp/slides/VisionSeminar14.pdf

**Comment by gar1t** [comment]
For simple object detection, we don't need semantic (pixel level) segmentation. This project might work just as well:

https://github.com/fizyr/keras-retinanet

**Comment by gar1t** [comment]
Another implementation:

https://github.com/ppwwyyxx/tensorpack/tree/master/examples/FasterRCNN

**Comment by gar1t** [comment]
@mbernico Absolutely, let's stick with Keras.

**Comment by gar1t** [comment]
This is a good step-by-step walk through object detection in general:

http://cv-tricks.com/object-detection/faster-r-cnn-yolo-ssd/

**Comment by mbernico** [comment]
I really like this implementation of YOLO v2.  https://github.com/experiencor/keras-yolo2 

Have seen it used recently for identifying parts of cars

**Comment by rajshah4** [comment]
Are we going with https://github.com/matterport/Mask_RCNN for the workshop?
I also like YOLO (but haven't tried the Keras version), but always feel like its a black sheep among the object detection models.

**Comment by mbernico** [comment]
I think that of the set of object localization models mask rcnn is probably the hardest to teach / most complex, but it's probably state of the art.  YOLO is simple(ish) to understand and SS multibox detector is in the middle somewhere (and my fav).  

I'm good with mask rcnn since we have the code for it, but I do think it will be difficult to teach.

**Comment by gar1t** [comment]
I've run across a couple of posts that have helped my understanding of this lineage of models (cited above but copied here):

- https://blog.athelas.com/a-brief-history-of-cnns-in-image-segmentation-from-r-cnn-to-mask-r-cnn-34ea83205de4

- http://cv-tricks.com/object-detection/faster-r-cnn-yolo-ssd/

In particular the first link post shows the history that led to Mask RCNN. With that lead up, the Mask model feels like a straight forward step. The one I'm not totally grokking is Faster as it introduces some nifty stages to the pipeline.

I think it would be helpful for students to see the progression:

    R-CNN -> Fast R-CNN -> Faster R-CNN -> Mask R-CNN

If we covered SSD and Yolo as a matter of completeness, that'd be fine - I think students should probably at least hear they exist.

As a matter of working code, I'm currently spending time with a number of models: Faster, Mask, and SSD. I'm working with TF's [object_detection framework](https://github.com/tensorflow/models/tree/master/research/object_detection) which uses TF Slim. I don't want to use that as teaching material as the Keras code is so much simpler! But the object_detection stuff is nice in that it's basically one interface into the entire ecosystem of detection models. Once I get something working end-to-end (this will include fine tuning on a custom dataset) I'll demo what I have. It sounds complicated, and it is atm, but it'll be simplified once I've gotten through it :)

All that to say that if we feel Mask is too much to present, we can dial it down - but I think the alternative should probably be Faster R-CNN as it seems the seminal architecture.

On that note, this repo comes up for Keras and Fast RCNN:

- https://github.com/yhenon/keras-frcnn

but it refers to RetinaNet:

- https://github.com/fizyr/keras-retinanet

I'm not sure where RetinaNet falls along the spectrum of these detectors architecturally.

I do have working code for the RetinaNet model - but it doesn't perform nearly as well as the Keras Mask R-CNN implementation. I wasn't paying close attention to the pre-training dataset or backbone used in either pipeline and I suspect that's the reason. Once I have the TF object_detection models working I'll go back and add both Mask R-CNN and the Retinanet Keras models to the mix. These will all have the same workflow interface via Guild and will let us compare performance across the board.
